plugin/plugins/kafka/kafka.go

```diff
@@ -0,0 +1,166 @@
+	TOPIC_PREFIX_PARAM = "topic_prefix"
+
+	// FLUSH_TIMEOUT_MS_PARAM is the timeout setting passed to the producer.Flush(timeoutMs)
+	FLUSH_TIMEOUT_MS_PARAM = "flush_timeout_ms"
+
+	// PRODUCER_CONFIG_PARAM is a map of the Kafka Producer configuration properties
+	PRODUCER_CONFIG_PARAM = "producer"
+
+	// KEYS_PARAM is a list of the StoreKeys we want to expose for this streaming service
+	KEYS_PARAM = "keys"
+
+	// HALT_APP_ON_DELIVERY_ERROR whether or not to halt the application when plugin fails to deliver message(s)
+	HALT_APP_ON_DELIVERY_ERROR = "halt_app_on_delivery_error"
+)
+
+// Plugins is the exported symbol for loading this plugin
+var Plugins = []plugin.Plugin{
+	&streamingServicePlugin{},
+}
+
+type streamingServicePlugin struct {
+	kss  *service.KafkaStreamingService
+	opts serverTypes.AppOptions
+}
+
+var _ plugin.StateStreamingPlugin = (*streamingServicePlugin)(nil)
+
+// Name satisfies the plugin.Plugin interface
+func (ssp *streamingServicePlugin) Name() string {
+	return PLUGIN_NAME
+}
+
+// Version satisfies the plugin.Plugin interface
+func (ssp *streamingServicePlugin) Version() string {
+	return PLUGIN_VERSION
+}
+
+// Init satisfies the plugin.Plugin interface
+func (ssp *streamingServicePlugin) Init(env serverTypes.AppOptions) error {
+	ssp.opts = env
+	return nil
+}
+
+// Register satisfies the plugin.StateStreamingPlugin interface
+func (ssp *streamingServicePlugin) Register(
+	bApp *baseapp.BaseApp,
+	marshaller codec.BinaryCodec,
+	keys map[string]*types.KVStoreKey,
+) error {
+	// load all the params required for this plugin from the provided AppOptions
+	tomlKeyPrefix := fmt.Sprintf("%s.%s.%s", plugin.PLUGINS_TOML_KEY, plugin.STREAMING_TOML_KEY, PLUGIN_NAME)
+	topicPrefix := cast.ToString(ssp.opts.Get(fmt.Sprintf("%s.%s", tomlKeyPrefix, TOPIC_PREFIX_PARAM)))
+	flushTimeoutMs := cast.ToInt(ssp.opts.Get(fmt.Sprintf("%s.%s", tomlKeyPrefix, FLUSH_TIMEOUT_MS_PARAM)))
+	haltAppOnDeliveryError := cast.ToBool(ssp.opts.Get(fmt.Sprintf("%s.%s", tomlKeyPrefix, HALT_APP_ON_DELIVERY_ERROR)))
+	producerConfig := cast.ToStringMap(ssp.opts.Get(fmt.Sprintf("%s.%s", tomlKeyPrefix, PRODUCER_CONFIG_PARAM)))
+	// get the store keys allowed to be exposed for this streaming service
+	exposeKeyStrings := cast.ToStringSlice(ssp.opts.Get(fmt.Sprintf("%s.%s", tomlKeyPrefix, KEYS_PARAM)))
+	var exposeStoreKeys []types.StoreKey
+
+	if len(exposeKeyStrings) > 0 {
+		exposeStoreKeys = make([]types.StoreKey, 0, len(exposeKeyStrings))
+		for _, keyStr := range exposeKeyStrings {
+			if storeKey, ok := keys[keyStr]; ok {
+				exposeStoreKeys = append(exposeStoreKeys, storeKey)
+			}
+		}
+	} else { // if none are specified, we expose all the keys
+		exposeStoreKeys = make([]types.StoreKey, 0, len(keys))
+		for _, storeKey := range keys {
+			exposeStoreKeys = append(exposeStoreKeys, storeKey)
+		}
+	}
+
+	// Validate minimum producer config properties
+	producerConfigKey := fmt.Sprintf("%s.%s.%s.%s", tomlKeyPrefix, PRODUCER_CONFIG_PARAM)
```

Comment:
The format has 4 inputs, but only 2 are provided.

Reply:
fixed